docs/iam-usage.rst

```diff
@@ -0,0 +1,177 @@
+
+>>> policy_change = iam.PolicyChange(version=2).add(iam.roles.OWNER.name, [iam.user('alice@example.com')])
+>>> policy, version, etag = policy_change.apply(resource)
+>>> print(policy)
+{'roles/owner': set(['user:alice@example.com'])}
+>>> print(version)
+2
+>>> print(etag)
+xDSFbfdasfAEFdfCds
+
+Apply returns the new policy, it's version, and it's etag.
+
+Optionally, a ``version`` keyword argument can be supplied to ``apply`` which will override the ``version`` behavior of the policy change.
+
+>>> _, version, _ = policy_change.apply(resource, version=None)
+>>> print(version)
+3
+
+Modifications can be added to a ``iam.PolicyChange`` object by one of two methods:
+
+First the user can directly add or remove members from a ``Role``. ``iam.PolicyChange`` exposes two methods for this, ``add`` and ``remove`` which both take a role string or ``iam.Role`` object and a list of member strings.
+
+>>> policy_change = iam.PolicyChange().add(iam.roles.OWNER, [iam.user('alice@example.com')])
+>>> policy_change.remove(iam.roles.EDITOR, [iam.domain('example.com'), iam.group('devs@example.com')])
+>>> policy, _, _ = policy_change.apply(resource)
+>>> print(iam.user('alice@example.com') in policy[iam.roles.OWNER.name])
+True
+>>> print(iam.domain('example.com') in policy[iam.roles.EDITOR.name])
+False
+
+Second the user can specify a "membership function" which will take a member string as an argument, and return ``True`` if the member should belong to the specified role, and ``False`` otherwise.
+
+>>> def membership_fn(member):
+>>>     return not iam.is_group(member) or member == iam.user('bob@example.com')
+>>> policy_change.fn(iam.roles.READER, membership_fn)
+>>> policy, _, _ policy_change.apply(resource)
+>>> print([member for member in policy[iam.roles.READER.name] if iam.is_group(member)])
+['user:bob@example.com']
+
+
+Methods
+----------------------------------
+
+Resources that implement the IAM interface provide the following methods:
+
+Low Level Methods
+~~~~~~~~~~~~~~~~~
+Resources that implement IAM provide low level methods for interacting with IAM. 
+
+``get_policy`` returns a tuple of ``(policy, version, etag)`` on the corresponding resource. 
+
+>>> policy, version, etag = resource.get_policy()
+>>> print(policy)
+{
+   'roles/owner': set(['user:alice@example.com']),
+   'roles/editor: set(['group:admins@example.com']),
+   'roles/reader': set(['domain:example.com', 'user:bob@example.com'])
+}
+>>> print(version)
+5
+>>> print(etag)
+ffdFADFdsgfsjrsHTY
+
+``set_policy`` takes a policy dictionary, as well as optional ``version`` and ``etag`` paramters. If updates are made to your policy during this change, they will be overwritten with exactly what is in your policy, or, if an etag is specified they will fail with a ``iam.ConcurrentModificationError``. ``iam.PolicyChange`` performs this "read-modify-write" cycle automatically for the user.
```

Comment:
Even with the retries, it's a fair amount of complexity. May be better off doing one of the following

- providing examples of how to retry-on-fail to let people build them into more complex apps
- ask the backend folks for support as-is
- ask the backend folks for transactions to avoid simulating them on the client